AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+/*
+ * Copyright (c) 2023, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/BuiltinWrappers.h>
+#include <AK/Span.h>
+#include <AK/StdLibExtras.h>
+#include <AK/Types.h>
+
+namespace AK {
+
+// Ideally, we want to store data in the native processor's words. However, for some algorithms,
+// particularly multiplication, we require double of the amount of the native word size.
+#if defined(__SIZEOF_INT128__) && defined(AK_ARCH_64_BIT)
+using NativeWord = u64;
+using DoubleWord = unsigned __int128;
+using SignedDoubleWord = __int128;
+#else
+using NativeWord = u32;
+using DoubleWord = u64;
+using SignedDoubleWord = i64;
+#endif
+
+template<bool sign>
+using ConditionallySignedDoubleWord = Conditional<sign, SignedDoubleWord, DoubleWord>;
+
+template<typename T>
+concept BuiltInUFixedInt = OneOf<T, bool, u8, u16, u32, u64, unsigned long, unsigned long long, DoubleWord>;
+
+template<typename T>
+constexpr inline size_t bit_width = sizeof(T) * 8;
+
+constexpr size_t word_size = bit_width<NativeWord>;
+constexpr NativeWord max_word = ~static_cast<NativeWord>(0);
+static_assert(word_size == 32 || word_size == 64);
```

Comment:
I'm not convinced that these types, constants and concepts need to be in the AK namespace rather than buried in Detail. Would anyone using the final BigInt classes need to access these declarations? And by "anyone" I mean LibCrypto, LibJS, etc.

Reply:
Don't think so.
Moved almost everything into AK::Detail. I left in AK only `StorageOperations`, `NativeWord`, `word_size`, `max_word`, `UnsignedStorageSpan`, `UnsignedStorageReadonlySpan` and `g_null_allocator` (previously `g_null_stack_bufer`). These things are enough even to make new BigInt class using BigIntBase.